lib/models/os.ts

```diff
@@ -0,0 +1,674 @@
+import first = require('lodash/first');
+import includes = require('lodash/includes');
+import once = require('lodash/once');
+import reject = require('lodash/reject');
+import * as memoizee from 'memoizee';
+import * as semver from 'semver';
+
+import {
+	deviceTypes as deviceTypesUtil,
+	findCallback,
+	isDevelopmentVersion,
+	notFoundResponse,
+	onlyIf,
+	treatAsMissingApplication,
+} from '../util';
+
+import { hupActionHelper } from '../util/device-actions/os-update/utils';
+
+import { BalenaRequestStreamResult } from '../../typings/balena-request';
+import {
+	DeviceType,
+	ImgConfigOptions,
+	OsUpdateVersions,
+	OsVersions,
+} from '../../typings/balena-sdk';
+import { InjectedDependenciesParam, InjectedOptionsParam } from '../balena';
+
+const BALENAOS_VERSION_REGEX = /v?\d+\.\d+\.\d+(\.rev\d+)?((\-|\+).+)?/;
+
+const DEVICE_TYPES_ENDPOINT_CACHING_INTERVAL = 10 * 60 * 1000; // 10 minutes
+
+const getOsModel = function(
+	deps: InjectedDependenciesParam,
+	opts: InjectedOptionsParam,
+) {
+	const { request, pubsub } = deps;
+	const { apiUrl, isBrowser } = opts;
+
+	const configModel = once(() => require('./config')(deps, opts));
+	const applicationModel = once(() => require('./application')(deps, opts));
+
+	// tslint:disable-next-line:ban-types
+	const withDeviceTypesEndpointCaching = <T extends Function>(fn: T) => {
+		const memoizedFn = memoizee(fn, {
+			maxAge: DEVICE_TYPES_ENDPOINT_CACHING_INTERVAL,
+			primitive: true,
+			promise: true,
+		});
+
+		pubsub.subscribe('auth.keyChange', () => memoizedFn.clear());
+
+		return memoizedFn;
+	};
+
+	/**
+	 * @summary Get device types with caching
+	 * @description Utility method exported for testability.
+	 * @name _getDeviceTypes
+	 * @private
+	 * @function
+	 * @memberof balena.models.os
+	 */
+	const _getDeviceTypes = withDeviceTypesEndpointCaching<
+		() => Promise<DeviceType[]>
+	>(() => configModel().getDeviceTypes());
+
+	const getValidatedDeviceType = (deviceTypeSlug: string) =>
+		_getDeviceTypes().then(types =>
+			deviceTypesUtil.getBySlug(types, deviceTypeSlug),
+		);
+
+	/**
+	 * @summary Get OS versions download size
+	 * @description Utility method exported for testability.
+	 * @name _getDownloadSize
+	 * @private
+	 * @function
+	 * @memberof balena.models.os
+	 */
+	const _getDownloadSize = withDeviceTypesEndpointCaching(
+		(deviceType: string, version: string) =>
+			request
+				.send({
+					method: 'GET',
+					url: `/device-types/v1/${deviceType}/images/${version}/download-size`,
+					baseUrl: apiUrl,
+				})
+				.get('body')
+				.get('size'),
+	);
+
+	/**
+	 * @summary Get OS versions
+	 * @description Utility method exported for testability.
+	 * @name _getOsVersions
+	 * @private
+	 * @function
+	 * @memberof balena.models.os
+	 */
+	const _getOsVersions = withDeviceTypesEndpointCaching((deviceType: string) =>
+		request
+			.send({
+				method: 'GET',
+				url: `/device-types/v1/${deviceType}/images`,
+				baseUrl: apiUrl,
```

Comment:
Let's add back the comment that explains why this is not omitting the token, so that we don't forget about it.
> optionally authenticated, so we send the token in all cases

Reply:
Hmm decaffeinate seems to have removed it, nice catch.